File: behavioral/dart/iterator_structured/custom_list.dart
import 'interfaces.dart';

class CustomList<T> implements Aggregate<T> {
  _Node<T>? _head;
  
  void add(T item) {
    final newNode = _Node(item);
    if (_head == null) {
      _head = newNode;
    } else {
      _Node<T>? current = _head;
      while (current!.next != null) {
        current = current.next;
      }
      current.next = newNode;
    }
  }
  
  @override
  Iterator<T> createIterator() {
    return ListIterator(_head);
  }
}

class _Node<T> {
  T data;
  _Node<T>? next;
  
  _Node(this.data);
}

class ListIterator<T> implements Iterator<T> {
  _Node<T>? _current;
  
  ListIterator(this._current);
  
  @override
  bool hasNext() {
    return _current != null;
  }
  
  @override
  T next() {
    if (!hasNext()) return null as T;
    final data = _current!.data;
    _current = _current!.next;
    return data;
  }
}

File: behavioral/typescript/iterator_structured/custom_list.ts
import { Iterator, Aggregate } from './interfaces';

export class CustomList<T> implements Aggregate<T> {
  private head: Node<T> | null = null;
  private count: number = 0;
  
  add(item: T): void {
    const newNode = new Node(item);
    if (!this.head) {
      this.head = newNode;
    } else {
      let current = this.head;
      while (current.next) {
        current = current.next;
      }
      current.next = newNode;
    }
    this.count++;
  }
  
  createIterator(): Iterator<T> {
    return new ListIterator(this.head);
  }
}

class Node<T> {
  data: T;
  next: Node<T> | null = null;
  
  constructor(data: T) {
    this.data = data;
  }
}

export class ListIterator<T> implements Iterator<T> {
  private current: Node<T> | null;
  
  constructor(head: Node<T> | null) {
    this.current = head;
  }
  
  hasNext(): boolean {
    return this.current !== null;
  }
  
  next(): T {
    if (!this.hasNext()) return null as T;
    const data = this.current!.data;
    this.current = this.current!.next;
    return data;
  }
}
